### PR TITLE
[HZ-1221] Assign WM keys, support keyed WMs in more processors

### PR DESCRIPTION
**Pull request by Fly-Style**
In https://github.com/hazelcast/hazelcast/pull/21575 we introduced keyed watermark mechanism which helps to determine the watermark input origin. But, some processors like `SessionWindowP`  and `SlidingWindowP` required additional tuning to fully support keyed watermarks. This PR introduces such support for processors mentioned above.

Also, this PR introduces the watermark keys assigner class, which automatically propagates watermark keys for optimized physical rel tree: from input relations to root relation. **Known limitation** : it doesn't support multiple watermarks, since Stream-to-Stream JOIN is not introduced yet -- the situation when the row has multiple watermarks is not possible -- no need to introduce additional complexity.

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+//
+//                assert leftRefByteMap.size() + rightRefByteMap.size() == jointRefByteMap.size();
+//                refToWmKeyMapping.put(join, jointRefByteMap);
+            } else if (node instanceof Join) {
+                // Hash Join and Nested Loop Join just forward watermarks from left input.
+                Join join = (Join) node;
+                Map<Integer, Byte> refByteMap = relToWmKeyMapping.get(join.getLeft());
+                relToWmKeyMapping.put(node, refByteMap);
+            } else {
+                // anything else -- just forward without any changes.
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:174]: Not true, we actually don't forward anything. The logic here has to match that in `HazelcastRelMdWatermarkedFields`

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+                    return;
+                }
+
+                int idx = 0;
+                Map<Integer, Byte> refByteMap = relToWmKeyMapping.get(union.getInput(idx++));
+                for (int i = idx; i < union.getInputs().size(); ++i) {
+                    relToWmKeyMapping.put(union.getInput(i), refByteMap);
+                }
+
+                RelVisitor topDownWmKeysAssigner = new TopDownUnionWatermarkKeyAssignerVisitor(
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java]: Instead of traversing the sub-tree, we can use some mutable reference and only change that reference.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                        int idx = ((RexInputRef) rexNode).getIndex();
+                        calcRefByteMap.put(calcIndex++, refByteMap.get(idx));
+                    }
+                }
+
+                relToWmKeyMapping.put(calc, calcRefByteMap);
+            } else if (node instanceof UnionPhysicalRel) {
+                UnionPhysicalRel union = (UnionPhysicalRel) node;
+                Set<Integer> usedColumns = new HashSet<>();
+                for (RelNode input : union.getInputs()) {
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:119]: We need to do intersection of watermarked fields. A field is watermarked if it's watermarked in all inputs.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+//                for (Map.Entry<RexInputRef, RexInputRef> entry : jointToRightInputMapping.entrySet()) {
+//                    if (rightRefByteMap.get(entry.getValue()) != null) {
+//                        jointRefByteMap.put(entry.getKey(), rightRefByteMap.get(entry.getValue()));
+//                    }
+//                }
+//
+//                assert leftRefByteMap.size() + rightRefByteMap.size() == jointRefByteMap.size();
+//                refToWmKeyMapping.put(join, jointRefByteMap);
+            } else if (node instanceof Join) {
+                // Hash Join and Nested Loop Join just forward watermarks from left input.
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:168]: This is not true for s2s join. It also needs to be fixed in `HazelcastRelMdWatermarkedFields`

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java (hunk 4 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:187]: S2S join rel doesn't exist in this PR, so, existing behavior is preserved. But, I prepared the code for S2S above (commented). I may remove it if you want.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+
+/**
+ * Traverse a relational tree and assign watermark keys.
+ */
+public class WatermarkKeysAssigner {
```
Comment (krzysztofslusarski) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:47]: Don't we need some unit test for that class?

**Review by krzysztofslusarski — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/kafka/RowProjectorProcessorSupplier.java [hunk 6]
```diff
@@ -51,8 +51,9 @@ final class RowProjectorProcessorSupplier implements ProcessorSupplier, DataSeri
 
     private Properties properties;
     private String topic;
-    private FunctionEx<ExpressionEvalContext, EventTimePolicy<JetSqlRow>> eventTimePolicyProvider;
+    private BiFunctionEx<ExpressionEvalContext, Byte, EventTimePolicy<JetSqlRow>> eventTimePolicyProvider;
     private KvRowProjector.Supplier projectorSupplier;
+    private byte watermarkKey;
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/kafka/RowProjectorProcessorSupplier.java:56]: It's needed, good catch re: ser/deser, ty!

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/utils/MutableByte.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt.utils;
```
[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/utils/MutableByte.java:17]: Not sure, tbh. It's our specific class for a specific reason...

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java (hunk 5 above)
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/WatermarkKeysAssigner.java:47]: We need, but it take some time...

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/utils/MutableByte.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (krzysztofslusarski) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/utils/MutableByte.java:17]: My motivation is that there is already ```MutableInteger``` and ```MutableLong``` in that package.

[the remark this answers is not in the document]
Comment (Fly-Style) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/utils/MutableByte.java:17]: Then, ok

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/FullScan.java [hunk 8]
```diff
@@ -32,25 +32,23 @@
...
     protected FullScan(
             RelOptCluster cluster,
             RelTraitSet traitSet,
             RelOptTable table,
-            @Nullable FunctionEx<ExpressionEvalContext, EventTimePolicy<JetSqlRow>> eventTimePolicyProvider,
+            @Nullable BiFunctionEx<ExpressionEvalContext, Byte, EventTimePolicy<JetSqlRow>> eventTimePolicyProvider,
             int watermarkedColumnIndex
     ) {
         super(cluster, traitSet, emptyList(), table);
-        assert watermarkedColumnIndex < 0 ^ eventTimePolicyProvider != null;
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/FullScan.java:46]: Why was this removed? I think it's still the case.

**Review by viliam-durina — CHANGES_REQUESTED** [review]

**Review by krzysztofslusarski — APPROVED** [review]

**Review by viliam-durina — APPROVED** [review]
